### PR TITLE
Spider class

### PR DESCRIPTION
**Pull request by byvernault**
This branch has been focused on the integration of a spider class for the spiders that we created. To follow the template of processor class / module class, we created three classes: one base class, one scan class, and one class for session.

With the new executable GenerateSpiderTemplate, you can generate in two seconds a nice python script with a proper name to start your spider script. You will need to fill the different pieces in the middle to obtain a fully functional spider.

Bug fixes:
    - fixing bug on logger printing to stderr. Now sending to stdout.

News:
    - adding time_writer to the SpiderProcessHandler (for the new spider template). By default, set to None and it should not change anything for the old spider.

Cheers,

Ben

### CODE REVIEW

**Comment by byvernault** [comment]
1) I have adding to this branch the jar files for XNAT to add the two datatypes used by DAX. The source code is present as well for people to edit if they want this datatypes. rebuild.sh will rebuild the jar files.
A tutorial has been written to explain how to install those files.

2) I have changed the code to process the host/user/pass with the spider. It's using the variable given to the spider class first, if not set for host and user, it's going to use the environment variable. If the env variable doesn't exist either, a ValueError will be raised. 
For the pass, if not set, it will look at the env variable if the user wasn't set, if the user is set, ask for the password on the terminal.

3) to the spider arguments, --host and --user have been added to allow the user to give those information to the spider command line.

Best,

Ben

File: bin/dax_tools/GenerateSpiderTemplate [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    f_obj.writelines(spider_code)
+    f_obj.close()
+
+def parse_args():
+    """
+        Parser for arguments
+    """
+    from argparse import ArgumentParser
+    usage = "Generates snapshots/preview for the scans for all subjects in a project."
+    argp = ArgumentParser(prog='make_preview_xnat', description=usage)
```
Comment (bud42) [inline bin/dax_tools/GenerateSpiderTemplate]: wrong name

File: dax/spiders.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+        # Xnat info:
+        self.xnat_project = xnat_project
+        self.xnat_subject = xnat_subject
+        self.xnat_session = xnat_session
+        # Xnat connection settings:
+        self.host = self.get_default_value("host", "XNAT_HOST", xnat_host)
+        self.user = self.get_default_value("user", "XNAT_USER", xnat_user)
+        self.pwd = self.get_pwd(xnat_pass, xnat_user)
+        # Suffix
+        if suffix[0] != '_': self.suffix = '_'+suffix
```
Comment (bud42) [inline dax/spiders.py]: what if suffix is empty string?

Comment (aplassard) [inline dax/spiders.py]: You could make it
if suffix and suffix[0] != '_'

if '' evaluates to false.

Comment (byvernault) [inline dax/spiders.py]: You are right. I need to add a conditions ig suffix not empty. I will test it tomorrow.

Thanks Brian,

Ben

File: bin/dax_tools/GenerateSpiderTemplate [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+                        #
+                        }}
+        self.upload_dict(results_dict)
+        self.end()
+
+if __name__ == '__main__':
+    args = parse_args()
+    # generate spider object:
+    spider_obj = Spider_{name}(spider_path=sys.argv[0],
+                               xnat_project=args.proj_label,
```
Comment (bud42) [inline bin/dax_tools/GenerateSpiderTemplate]: we are missing the jobdir argument here

Comment (byvernault) [inline bin/dax_tools/GenerateSpiderTemplate]: ok, I think I fixed both issues.

Thanks for reporting them Brian.

Let me know if you anymore of them.

Cheers,

Ben

File: bin/dax_tools/GenerateProcessorTemplate [hunk 4]
```diff
@@ -164,7 +163,6 @@ class {name}_Processor(SessionProcessor):
                  spm_path=PathSettings.SPM_PATH, suffix_proc=''):
```
Comment (bud42) [inline bin/dax_tools/GenerateProcessorTemplate]: what is PathSettings?

File: bin/dax_tools/GenerateProcessorTemplate [hunk 5]
```diff
@@ -141,8 +142,8 @@ class Processor_{name}(ScanProcessor):
 
 SESSION_LEVEL_TEMPLATE = DEFAULT_TEMPLATE+"""
 # Format for the spider command line
-SPIDER_FORMAT = '''python {{spider}} -v --spm {{spm}} \
--p {{proj}} -s {{subj}} -e {{sess}} -d {{dir}} --suffix "{{suffix_proc}}"'''
+SPIDER_FORMAT = '''python {{spider}} -p {{proj}} -s {{subj}} -e {{sess}}\
```
Comment (bud42) [inline bin/dax_tools/GenerateProcessorTemplate]: need a space before the \

Comment (byvernault) [inline bin/dax_tools/GenerateProcessorTemplate]: Yes for the space.

You can remove the pathsettings. I thought I remove all of them.

Ben

Comment (aplassard) [inline bin/dax_tools/GenerateProcessorTemplate]: I'm just curious and I would check this if I was in front of a computer,
but how is path settings handled? Is it part of the dax package or is it a
python settings file like the grid engine file?

Comment (byvernault) [inline bin/dax_tools/GenerateProcessorTemplate]: It was when I created this for the dax-spider package. It should not be there. 

Ben

**Pull request merged**